toonz/sources/toonzlib/toonzscene.cpp

```diff
@@ -169,71 +169,34 @@ void deleteUntitledScene(const TFilePath &fp) {
 }
 
 //-----------------------------------------------------------------------------
-/*-- TODO: オプション化して復活させるか、検討のこと --*/
-/*
-void saveBackup(const TFilePath &fp)
-{
-  if(!TFileStatus(fp).doesExist()) return;
-        wstring sceneName = fp.getWideName();
-  TFilePath bckDir = fp.getParentDir() + "backups" + sceneName;
-  if(!TFileStatus(bckDir).doesExist())
-  {
-    try {TSystem::mkDir(bckDir);}
-    catch(...) {return;}
-  }
-
-  std::map<int, TFilePath> oldBackups;
-  TFilePathSet lst = TSystem::readDirectory(bckDir);
-  for(TFilePathSet::iterator it = lst.begin(); it != lst.end(); ++it)
-  {
-    TFilePath fp2 = *it;
-    if(fp2.getType() != "tnz" && fp2.getType() != "tab") continue;
-    wstring name = fp2.getWideName();
-    if(name.find_first_of(L"0123456789") == wstring::npos) continue;
-                int i = name.find(sceneName);
-                if(i != wstring::npos)
-                        name = name.substr(sceneName.size()+1);
-    if(name == L"" || name.find_first_not_of(L"0123456789") != wstring::npos)
-      continue;
-    int index = toInt(name);
-    assert(0<index);
-    assert(oldBackups.count(index)==0);
-    oldBackups[index] = fp2;
-  }
 
-  int m = 3;
-  if((int)oldBackups.size()>m)
-  {
-    std::map<int, TFilePath>::iterator it = oldBackups.begin();
-    for(int i=0;i+m<(int)oldBackups.size();i++)
-    {
-      assert(it != oldBackups.end());
-      TFilePath toKill = it->second;
-      try {TSystem::deleteFile(toKill); } catch(...) {}
-      ++it;
+static void saveBackup(TFilePath path) {
+  int totalBackups = Preferences::instance()->getBackupKeepCount();
+  totalBackups -= 1;
+  TFilePath backup = path.withType(path.getType() + ".bak");
+  TFilePath prevBackup =
+      path.withType(path.getType() + ".bak" + std::to_string(totalBackups));
+  while (--totalBackups >= 0) {
+    std::string bakExt =
+        ".bak" + (totalBackups > 0 ? std::to_string(totalBackups) : "");
+    backup = path.withType(path.getType() + bakExt);
+    if (!TSystem::doesExistFileOrLevel(backup)) continue;
+    try {
+      if (TSystem::doesExistFileOrLevel(prevBackup))
+        TSystem::removeFileOrLevel_throw(prevBackup);
```

Comment:
I think you don't need to check existence of `prevBackup` and remove it here, because the subsequent function `TSystem::copyFileOrLevel_throw` will do the overwrite copy.

The same goes for the function in `txshsimplelevel.cpp` .

Reply:
That logic mimics the logic that was already there in the original saveBackup logic. I can certainly remove it.